trio/_subprocess/unix_pipes.py

```diff
@@ -0,0 +1,119 @@
+import os
+from typing import Tuple
+
+from .. import _core, BrokenStreamError
+from .._abc import SendStream, ReceiveStream
+
+__all__ = ["PipeSendStream", "PipeReceiveStream", "make_pipe"]
+
+
+class _PipeMixin:
+    def __init__(self, pipefd: int, *, set_non_blocking: bool = True):
+        if not isinstance(pipefd, int):
+            raise TypeError(
+                "{0.__class__.__name__} needs a pipe fd".format(self)
+            )
+
+        self._pipe = pipefd
+        self._closed = False
+
+        if set_non_blocking:
+            import fcntl
+            flags = fcntl.fcntl(self._pipe, fcntl.F_GETFL)
+            fcntl.fcntl(self._pipe, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
+    async def aclose(self):
+        if self._closed:
+            return
+
+        os.close(self._pipe)
+        _core.notify_fd_close(self._pipe)
+        self._closed = True
+        await _core.checkpoint()
+
+    def fileno(self) -> int:
+        """Gets the file descriptor for this pipe."""
+        return self._pipe
+
+    def __del__(self):
+        if self._closed:
+            return
+
+        try:
+            os.close(self._pipe)
+        except AttributeError:
+            # probably in interpreter shut down
+            pass
+        except OSError as e:
+            # already closed from somewhere else
+            if e.errno != 9:
```

Comment:
Whoops, you should never hard-code errno values like this – they're arbitrary, and can be different on different platforms. (Though 9 does seem to be EBADF on both Linux and macOS... weird!) Instead, use the stdlib `errno` module, e.g. `e.errno != errno.EBADF`.

Except... in this case, I actually think we shouldn't be filtering out EBADF. Instead we should follow the rule that once you give an fd to one of these pipe objects, the pipe object "takes ownership" of it, and is the only object that should be used to close it. Closing the fileno somewhere else is a bug.

In fact, this probably explains some of your weird test failures. I bet what's happening is:

1. The test allocates the fd, and wraps it in a trio pipe object
2. The test closes the fd by hand
3. Later, some other fd is opened, and is assigned the same numerical value as the earlier pipe fd.
4. Finally, the trio pipe object gets GCed, and its `__del__` method closes whatever random fd was opened in step 3.
5. Chaos ensues.

File descriptors are a weird kind of global state with no guard-rails. They're very C-ish. It's important to know exactly who is responsible for managing them.